### PR TITLE
In checkpoints, serialize hash of attachment instead of its data

### PR DESCRIPTION
**Pull request by andr3ej**
This PR implements https://r3-cev.atlassian.net/browse/CORDA-300

### CODE REVIEW

**Comment by mikehearn** [comment]
I'm not totally sure why a new interface is required? Can't we just change the attachment storage so it does its own dbtx? or, run checkpoint deserialisation inside a transaction in the SMM?

Using the serialize-as-token infrastructure might not be the best way to do this, if it has to be extended. It's really meant for singleton classes. Not something like an attachment.

**Comment by rick-r3** [comment]
There should be a database transaction _in-flight_ when restoring the checkpoints since they themselves are read from the database, so.... agree we shouldn't need a new interface (just use `AttachmentStorage`) or to open any new database transactions.

Well done for finding `SerializeAsToken` but, as Mike is saying, it wasn't designed for this since it assumes a finite set of tokens (see the presence of an internal map) which isn't the case here.  I think you'll end up having to write a custom serialiser in `Kryo.kt`.

**Comment by andr3ej** [comment]
Changes since the last version of this PR:
* Quasar won't let us have a reference to the db at deserialisation time (it clears all ThreadLocals) so attachments now reload their data lazily
* SerializeAsToken with custom tokens appears to do exactly what I want so I'm not keen to reimplement that, it's only SingletonSerializationToken that backs itself by the singleton map...
* ... which I've refactored so we can get singletons out by supertype, which AbstractAttachment does to get the StorageService for reloading attachment data (practical motivation is it can't see the StorageServiceImpl class)

File: core/src/main/kotlin/net/corda/core/contracts/Structures.kt [hunk 1]
```diff
@@ -513,3 +512,21 @@ interface Attachment : NamedByHash {
         throw FileNotFoundException(path)
     }
 }
+
+abstract class AbstractAttachment(dataLoader: () -> ByteArray) : Attachment {
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/contracts/Structures.kt]: Is this really meant to be in the core module? I'd think it'd be an implementation detail of the node.

Comment (andr3ej) [inline core/src/main/kotlin/net/corda/core/contracts/Structures.kt]: It's used by FetchAttachmentsFlow which is in core

File: core/src/main/kotlin/net/corda/core/serialization/SerializationToken.kt [hunk 2]
```diff
@@ -75,14 +82,14 @@ class SerializeAsTokenSerializer<T : SerializeAsToken> : Serializer<T>() {
  * on the Kryo instance when serializing to enable/disable tokenization.
  */
 class SerializeAsTokenContext(toBeTokenized: Any, kryoPool: KryoPool) {
-    internal val tokenToTokenized = mutableMapOf<SerializationToken, SerializeAsToken>()
-    internal var readOnly = false
+    private val classNameToSingletons = Multimaps.newMultimap<String, SerializeAsToken>(HashMap(), { newSetFromMap(IdentityHashMap()) })
```
Comment (andr3ej) [inline core/src/main/kotlin/net/corda/core/serialization/SerializationToken.kt]: Yes, the changes are essentially 2 things:
* Add uniqueSingletonImplementing and related data structure changes
* Code moved from the token to the context as more things can be private that way

I think making the SATC more specific would work, but:
* That would be a bigger-impact change for another PR, I haven't actually changed what's stored in the SATC in this PR, only how it's stored
* This approach (lookup by type, analogous to constructor injection) means you don't have to remember a chain of field names to get to the object you want, see Law of Demeter
* I'm preparing a PR to remove Kryo.attachmentStorage that will take advantage of this facility

File: core/src/main/kotlin/net/corda/core/serialization/SerializationToken.kt [hunk 3]
```diff
@@ -95,30 +102,44 @@ class SerializeAsTokenContext(toBeTokenized: Any, kryoPool: KryoPool) {
...
+    internal fun uniqueSingleton(className: String, exactMatch: Boolean): Any {
+        val singletons = classNameToSingletons[className]
+        if (1 != singletons.size) noSuchSingleton(className)
+        val singleton = singletons.first()
+        if (exactMatch && singleton.javaClass.name != className) noSuchSingleton(className)
+        return singleton
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    fun <T : Any> uniqueSingletonImplementing(clazz: KClass<T>) = uniqueSingleton(clazz.java.name, false) as T
```
Comment (andr3ej) [inline core/src/main/kotlin/net/corda/core/serialization/SerializationToken.kt]: Done

File: core/src/main/kotlin/net/corda/core/serialization/SerializationToken.kt (hunk 2 above)
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/serialization/SerializationToken.kt]: I think adding `ServiceHub` to SATC is a smaller change.  It's just the constructor needs an extra `val services: ServiceHub`, and then change the single call site (outside tests).   No other changes to STAC would then be required. Whilst you may have a point re: Law of Demeter I'd defer that to how `ServiceHub` looks and is implemented, although personally I prefer my services to be discoverable via code complete in the IDE...

Comment (andr3ej) [inline core/src/main/kotlin/net/corda/core/serialization/SerializationToken.kt]: Done

File: core/src/main/kotlin/net/corda/flows/FetchAttachmentsFlow.kt [hunk 4]
```diff
@@ -15,19 +18,21 @@ class FetchAttachmentsFlow(requests: Set<SecureHash>,
...
         }
     }
 
-    private class ByteArrayAttachment(private val wire: ByteArray) : Attachment {
-        override val id: SecureHash by lazy { wire.sha256() }
-        override fun open(): InputStream = wire.inputStream()
-        override fun equals(other: Any?) = other === this || other is Attachment && other.id == this.id
-        override fun hashCode(): Int = id.hashCode()
-        override fun toString(): String = "${javaClass.simpleName}(id=$id)"
+    private class ByteArrayAttachment(dataLoader: () -> ByteArray) : AbstractAttachment(dataLoader), SerializeAsToken {
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/flows/FetchAttachmentsFlow.kt]: I wonder if this should be named `ImportedAttachment` or `PersistedAttachment` or something like that?  `ByteArrayAttachment` sounds like it is standalone, but in fact for the tokenisation to work it must be already persisted (which it always is for this flow).

File: core/src/main/kotlin/net/corda/core/serialization/SerializationToken.kt [hunk 5]
```diff
@@ -74,15 +76,15 @@ class SerializeAsTokenSerializer<T : SerializeAsToken> : Serializer<T>() {
  * Then it is a case of using the companion object methods on [SerializeAsTokenSerializer] to set and clear context as necessary
  * on the Kryo instance when serializing to enable/disable tokenization.
  */
-class SerializeAsTokenContext(toBeTokenized: Any, kryoPool: KryoPool) {
-    internal val tokenToTokenized = mutableMapOf<SerializationToken, SerializeAsToken>()
-    internal var readOnly = false
+class SerializeAsTokenContext(toBeTokenized: Any, kryoPool: KryoPool, val serviceHub: ServiceHub) {
+    private val classNameToSingleton = mutableMapOf<String, SerializeAsToken>()
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/serialization/SerializationToken.kt]: We could have left this alone, but agree this is a fair change to discourage accidental caching of non-singletons that might bloat out the map, so let's go with it.

File: core/src/test/kotlin/net/corda/core/serialization/AttachmentSerializationTest.kt [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+            update(AttachmentEntity().apply {
+                attId = attachmentId
+                content = data
+            })
+        }
+    }
+}
+
+class AttachmentSerializationTest {
+
```
Comment (rick-r3) [inline core/src/test/kotlin/net/corda/core/serialization/AttachmentSerializationTest.kt]: Remove blank line

File: core/src/test/kotlin/net/corda/core/serialization/AttachmentSerializationTest.kt [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+        override fun call() {
+            receive<String>(client).unwrap { assertEquals("ping one", it) }
+            sendAndReceive<String>(client, "pong").unwrap { assertEquals("ping two", it) }
+        }
+    }
+
+    private class ClientResult(internal val attachmentContent: String)
+
+    private abstract class ClientLogic(server: MockNetwork.MockNode) : FlowLogic<ClientResult>() {
+
```
Comment (rick-r3) [inline core/src/test/kotlin/net/corda/core/serialization/AttachmentSerializationTest.kt]: Remove blank line

File: core/src/test/kotlin/net/corda/core/serialization/AttachmentSerializationTest.kt [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+            sendAndReceive<String>(server, "ping one").unwrap { assertEquals("pong", it) }
+            send(server, "ping two")
+        }
+
+        @Suspendable
+        override fun call() = ClientResult(getAttachmentContent())
+
+        @Suspendable // This annotation is required by the instrumentation verifier.
+        internal abstract fun getAttachmentContent(): String
+
```
Comment (rick-r3) [inline core/src/test/kotlin/net/corda/core/serialization/AttachmentSerializationTest.kt]: Remove blank line

File: core/src/test/kotlin/net/corda/core/serialization/AttachmentSerializationTest.kt [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+    }
+
+    @Test
+    fun `only the hash of a FetchAttachmentsFlow attachment should be saved in checkpoint`() {
+        val attachmentId = server.saveAttachment("genuine")
+        launchFlow(FetchAttachmentLogic(server, attachmentId), 2)
+        client.hackAttachment(attachmentId, "hacked")
+        assertEquals("hacked", rebootClientAndGetAttachmentContent(false))
+    }
+
```
Comment (rick-r3) [inline core/src/test/kotlin/net/corda/core/serialization/AttachmentSerializationTest.kt]: Remove blank line

**Comment by andr3ej** [comment]
Changes made

**Pull request merged**